### PR TITLE
Eliminate alien nodes before execution for entry db

### PR DESCRIPTION
**Pull request by seryozhasmirnov**
https://github.com/greenplum-db/gpdb/pull/2588 eliminates alien nodes before execution but entry db nodes was missed.
This PR adds alien nodes elimination for entry db.
There is a small fix in `formIdleSegmentIdList` function to  propagate GUC change to **entry db** (e.g. `execute_pruned_plan` that was used in test).

The is a PR https://github.com/arenadata/gpdb/pull/352 in GPDB fork in **6x** that also adds test for entry db nodes elimination.

As there is no memory accounting for **X_Alien** nodes in **master** alien nodes elimination can be tested manually (i.e. test from https://github.com/greenplum-db/gpdb/pull/2663 is not applicable).

### CODE REVIEW

**Comment by kainwen** [comment]
Looks good at first glance. Thanks for your contribution.

Please kindly wait for more time and let me and the team look into this.

File: src/backend/cdb/dispatcher/cdbdisp_query.c [hunk 1]
```diff
@@ -1423,6 +1423,16 @@ formIdleSegmentIdList(void)
 		}
 	}
 
+	if (cdbs->entry_db_info != NULL)
```
Comment (zxuejing) [inline src/backend/cdb/dispatcher/cdbdisp_query.c:1426]: hi @seryozhasmirnov, thanks for your contribution.
I have a question. 
These codes are not related to eliminating alien nodes for entry db.
this is related to set command, Is that right?

File: src/backend/executor/execMain.c [hunk 2]
```diff
@@ -557,7 +557,7 @@ standard_ExecutorStart(QueryDesc *queryDesc, int eflags)
 	 *
 	 * TODO: eliminate aliens even on master, if not EXPLAIN ANALYZE
```
Comment (zxuejing) [inline src/backend/executor/execMain.c:558]: Todo should be removed.

File: src/backend/cdb/dispatcher/cdbdisp_query.c (hunk 1 above)
Comment (kainwen) [reply src/backend/cdb/dispatcher/cdbdisp_query.c:1426]: > hi @seryozhasmirnov, thanks for your contribution. I have a question. These codes are not related to eliminating alien nodes for entry db. this is related to set command, Is that right?

Ah, good finding. I miss this part. It seems that this function `formIdleSegmentIdList` is only invoked during `Set` statement.

File: src/backend/executor/execMain.c (hunk 2 above)
Comment (kainwen) [reply src/backend/executor/execMain.c:558]: > Todo should be removed.

I think the TODO here `on master` really means `on QD`, so we might not remove it.

File: src/backend/cdb/dispatcher/cdbdisp_query.c (hunk 1 above)
Comment (seryozhasmirnov) [reply src/backend/cdb/dispatcher/cdbdisp_query.c:1426]: Hi @zxuejing and @kainwen,

I have found that  `execute_pruned_plan` GUC is not changed for **entry db**. There is a small fix in `formIdleSegmentIdList` function.

The simple test to check alien nodes elimination:

```
create table t1 (a int, b int, c int) distributed by (a);
\pset pager off
set explain_memory_verbosity=detail;
set execute_pruned_plan to on;
explain analyze select * from t1, (select * from t1 limit 1) a where a.a = t1.a;
set execute_pruned_plan to off;
explain analyze select * from t1, (select * from t1 limit 1) a where a.a = t1.a;
```

With `execute_pruned_plan` GUC switched to `on` memory consumption for **X_Alien** nodes is 0, with `execute_pruned_plan` GUC switched to `off` memory consumption for **X_Alien** nodes exceeds 0.
Without fix in `formIdleSegmentIdList` function memory consumption for **X_Alien** nodes remains 0 for **entry db** as `execute_pruned_plan` GUC remains `on` despite changing it to `off`.

File: src/backend/executor/execMain.c (hunk 2 above)
Comment (seryozhasmirnov) [reply src/backend/executor/execMain.c:558]: Dear @zxuejing and @kainwen,
So keep TODO or remove?

Comment (zxuejing) [reply src/backend/executor/execMain.c:558]: We can keep TODO to record we do not eliminate aliens even on master(QD)

File: src/backend/cdb/dispatcher/cdbdisp_query.c (hunk 1 above)
Comment (zxuejing) [reply src/backend/cdb/dispatcher/cdbdisp_query.c:1426]: hi, @seryozhasmirnov  please add some comments here.
Others are LGTM.

Comment (zxuejing) [reply src/backend/cdb/dispatcher/cdbdisp_query.c:1426]: Add a comment to PR that is enough.

**Review by zxuejing — APPROVED** [review]

**Comment by ashwinstar** [comment]
@zxuejing @kainwen if all the comments of the PR is addressed, please can we merge the same?

**Comment by kainwen** [comment]
> @zxuejing @kainwen if all the comments of the PR is addressed, please can we merge the same?

Last night xuejing run a full pipeline for this and decide to push it today based on the ci result.
